var/spack/repos/builtin/packages/py-cupy/package.py

```diff
@@ -17,13 +17,23 @@ class PyCupy(PythonPackage, CudaPackage):
     pypi = "cupy/cupy-8.0.0.tar.gz"
 
     version("11.2.0", sha256="c33361f117a347a63f6996ea97446d17f1c038f1a1f533e502464235076923e2")
-
-    depends_on("python@3.7:", type=("build", "run"))
+    version("11.3.0", sha256="d057cc2f73ecca06fae8b9c270d9e14116203abfd211a704810cc50a453b4c9e")
+    version("11.4.0", sha256="03d52b2626e02a3a2b46d714c1cd03e702c8fe33915fcca6ed8de5c539964f49")
+    version("11.5.0", sha256="4bc8565bded22cc89b210fd9fb48a5d5316f30701e12bb23852a60314e1f9f6e")
+    version("11.6.0", sha256="53dbb840072bb32d4bfbaa6bfa072365a30c98b1fcd1f43e48969071ad98f1a7")
+    version("12.0.0", sha256="61ddbbef73d50d606bd5087570645f3c91ec9176c2566784c1d486d6a3404545")
+    version("12.1.0", sha256="f6d31989cdb2d96581da12822e28b102f29e254427195c2017eac327869b7320")
+
+    depends_on("python@3.7:", when="@:11", type=("build", "run"))
+    depends_on("python@3.8:", when="@12:", type=("build", "run"))
     depends_on("py-setuptools", type="build")
     depends_on("py-cython@0.29.22:2", type="build")
     depends_on("py-fastrlock@0.5:", type=("build", "run"))
-    depends_on("py-numpy@1.20:1.25", type=("build", "run"))
-    depends_on("cuda")
+    depends_on("py-numpy@1.20:1.25", when="@:11", type=("build", "run"))
+    depends_on("py-numpy@1.20:1.26", when="@12:", type=("build", "run"))
+    depends_on("py-scipy@1.6:1.12", type=("build", "run"))
```

Comment:
Scipy is optional, not required. If you want you can add a `+all` variant to control whether or not scipy/cython/optuna get installed, but I would default to False so as to match the behavior of `pip install cupy`.

Reply:
Addressing your review remarks in follow up PR here:
* https://github.com/spack/spack/pull/38919